bfvmm/include/hve/arch/intel_x64/vcpu.h

```diff
@@ -258,6 +258,20 @@ class vcpu : public eapis::intel_x64::vcpu
     ///
     VIRTUAL uint64_t lapic_base() const;
 
+    /// Read
+    ///
+    /// @param indx the dword offset to read from
+    /// @return the 32-bit value of the register
+    ///
+    VIRTUAL uint32_t lapic_read(uint32_t indx) const;

```

Comment:
Not a big fan of these. Why would someone using the vCPU need raw access to APIC registers? I'd prefer that if you need to know something that is in the vCPU that we provide an API to get it which removes the train wreck.

Reply:
When APIC accesses are being handled from an exit, you are working with an indx. If we had explicity APIs for each register, we would have giant switch statements in the exit handlers. That being said, I agree that things external to the VCPU shouldn't need this; this is more for cllients within the vcpu i.e. xen_op_handler. So we could use a different pattern like a friend or something

Reply:
Agreed. The APIC class should be self-contained, included decoding instructions, etc... and the ept_violation handler should be in the APIC class itself. The APIC class should then expose whatever APIs the other classes need like setting the TSC mode, etc... 

The Xen-ops class just handles the resources for now, and the APIC is not part of that.